### PR TITLE
Hub api call tests

### PR DESCRIPTION
**Pull request by MAsimSENG**
Parent Trello Ticket associated with this PR: https://trello.com/c/ywH2J3bl/39-disabling-hubspot-connection

Two tests are created for the HubApiCall function (PR#21 https://github.com/SENG499-team-2/SaaSquatch-HubSpot-integration/pull/21) for more info about the function. 

The tests are based on the feature file: Disconnect.feature. 

The following two scenarios are tested:

1)Disconnecting Hubspot from integration (access token expired) 
2)Disconnecting Hubspot from integration and access token is not expired yet

The following line code is added to prevent CORS errors:
axios.defaults.adapter = require('axios/lib/adapters/http')

Consideration was given to use mock API calls using jest mock, however due to some errors faced and the nature the HubApiCall func real API calls were used.  In the second scenario,  an API call is made to an irrelevant URL which returns success.

### CODE REVIEW

File: src/routes/oath.js [hunk 1]
```diff
@@ -0,0 +1,280 @@
+"use strict";
```
Comment (davidbishoppp) [inline src/routes/oath.js]: add js files from src to gitignore.

File: src/test/HubApiCall_Disconnect.steps.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            // the api call will happen in the final then
+        });
+
+        and('the integration attempts to get a new access token from Hubspot', () => {
+            // this step is done in step a call to hubspotAPi function in the then below
+        });
+
+        then('a bad request error is returned', async () => {
+            const res = await HubApiCall(errorCall,refresh_token);
+            try {
```
Comment (davidbishoppp) [inline src/test/HubApiCall_Disconnect.steps.ts:69]: Does this need a try catch?

File: src/test/HubApiCall_Disconnect.steps.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+        when('I disconnect the integration from my Hubspot account', () => {
+            // this implies that the refresh token has been invalidated
+            // unable to get selenium working as of yet but for testing
+            // purposes we can use an invalid refresh token (random string)
+            refresh_token = 'invalid token example';
+        });
+
+        and('the current access token expires', () => {
+            //
```
Comment (davidbishoppp) [inline src/test/HubApiCall_Disconnect.steps.ts]: empty comment

File: src/test/HubApiCall_Disconnect.steps.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            access_token = "nowexpiredaccess_token";
+        });
+
+        then('requests from the integration to protected endpoints in Hubspot fail', () => {
+            // the api call will happen in the final then
```
Comment (davidbishoppp) [inline src/test/HubApiCall_Disconnect.steps.ts]: Make a standard request here without the HubApiCall function to show it will fail?

**Review by davidbishoppp — CHANGES_REQUESTED** [review]

File: src/routes/oath.ts [hunk 5]
```diff
@@ -150,19 +151,24 @@ export const HubApiCall:any = async function (myapifunc:Function,refresh_token:s
...
+
+
 		// api call went through everything is fine
 		return result;
 	}
 	catch(e)
 	{
 		try
 		{
+
```
Comment (tolkamps1) [inline src/routes/oath.ts]: Can we remove the extra spacing between lines?

File: src/test/HubApiCall_Disconnect.steps.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        });
+
+        then('requests from the integration to protected endpoints in Hubspot pass', () => {
+            // the api call will happen in the final then
+        });
+
+        and('a status 200 is returned',async () => {
+                const res = await HubApiCall(correctCall,refresh_token);
+
+                expect(res.status).toBe('success');
```
Comment (tolkamps1) [inline src/test/HubApiCall_Disconnect.steps.ts:128]: Suggestion: remove extra line breaks here. I think one empty line is sufficient between ending curly braces.

File: src/test/HubApiCall_Disconnect.steps.ts [hunk 7]
```diff
@@ -0,0 +1,135 @@
+import {defineFeature, loadFeature} from "jest-cucumber";
+import {HubApiCall} from "../routes/oath";
+const axios = require("axios");
+const Disconnect = loadFeature('features/Disconnect.feature');
+const querystring = require('query-string');
+axios.defaults.adapter = require('axios/lib/adapters/http')
+const correctCall = async () => {
+    try {
+        // this represents any api call that doesn't fail
+       const response = await axios.get('https://dog.ceo/api/breeds/list/all')
```
Comment (tolkamps1) [inline src/test/HubApiCall_Disconnect.steps.ts:13]: I'm a little confused on to why we need this. I know you are using it to check a 200 response status below, but confirming that this API call is working doesn't actually test our code. Can you explain more of why we have this?

**Review by tolkamps1 — CHANGES_REQUESTED** [review]

File: src/test/HubApiCall_Disconnect.steps.ts (hunk 4 above)
Comment (MAsimSENG) [reply src/test/HubApiCall_Disconnect.steps.ts:71]: For some reason if I make any api call from here I get this error: 
image

File: src/test/HubApiCall_Disconnect.steps.ts (hunk 7 above)
Comment (MAsimSENG) [reply src/test/HubApiCall_Disconnect.steps.ts:13]: This tests the part of HubApiCall shown in the image below: Ie the case where the access token hasn't expired so the API call is successful so it doesn't hit the catch part of the HubApiCall function.  I have added a to-do comment to update the API call to some HubSpot endpoint once there is a test account access token available in DB although that really won't change the overall goal of the test just will look better. 

(In oath.ts)
image

(in test file)

image

File: src/test/HubApiCall_Disconnect.steps.ts (hunk 2 above)
Comment (MAsimSENG) [reply src/test/HubApiCall_Disconnect.steps.ts:69]: I have removed it its not necessary

File: src/test/HubApiCall_Disconnect.steps.ts (hunk 4 above)
Comment (MAsimSENG) [reply src/test/HubApiCall_Disconnect.steps.ts]: Added some code as suggested

**Comment by MAsimSENG** [comment]
Made changes as requested, waiting for final satisfaction before merging.

**Comment by MAsimSENG** [comment]
[Disabling Hubspot connection](https://trello.com/c/ywH2J3bl/39-disabling-hubspot-connection)

File: src/routes/oath.js (hunk 1 above)
Comment (davidbishoppp) [reply src/routes/oath.js]: Can you make sure to add js files to the gitignore instead of deleting them

**Review by tolkamps1 — APPROVED** [review]

**Review by lucianojd — APPROVED** [review]
Everything looks good to me.

**Review by davidbishoppp — APPROVED** [review]

**Comment by MAsimSENG** [comment]
Added src/**/*.js to gitignore going to merge pr now.

**Pull request merged**